_drafts/zio-fibers.md

```diff
@@ -0,0 +1,290 @@
+
+## 3. ZIO Effect
+
+The ZIO library is one of the implementations of the effect pattern available in Scala. Its primary effect type is `ZIO[R, E, A]`. We call the `R` type the _environment type_, representing the dependencies the effect needs to execute. Once executed, the effect can produce a result of type `A` or fail, producing a value of the type `E`. For these reasons, it's common to think about the type`ZIO[R, E, A]` as an _effectful_ version of the function type `R => Either[E, A]`.
+
+**The default model of execution of the `ZIO` effect is synchronous**. To start showing some code, let's model some valuable scenarios. So, we will model the wake-up routine of Bob. Every morning, Bob wakes up, goes to the bathroom, then boils some water to finally prepare a cup of coffee:
+
+```scala
+val bathTime = ZIO.succeed("Going to the bathroom")
+val boilingWater = ZIO.succeed("Boiling some water")
+val preparingCoffee = ZIO.succeed("Preparing the coffee")
+```
+
+As the three effects cannot fail, we use the smart constructor `ZIO.succeed`. Moreover, to help us debug the execution thread of each effect, we define the following function:
+
+```scala
+def printThread = s"[${Thread.currentThread().getName}]"
+```
+
+With these bullets in our functional gun, we can compose the above effects and execute them:
+
+```scala
+import zio._
+
+object FibersTutorial extends zio.App {
+
+  def printThread = s"[${Thread.currentThread().getName}]"
+
+  val bathTime = ZIO.succeed("Going to the bathroom")
+  val boilingWater = ZIO.succeed("Boiling some water")
+  val preparingCoffee = ZIO.succeed("Preparing the coffee")
+
+  def sequentialWakeUpRoutine(): ZIO[Any, Nothing, Unit] = for {
+    _ <- bathTime.debug(printThread)
+    _ <- boilingWater.debug(printThread)
+    _ <- preparingCoffee.debug(printThread)
+  } yield ()
+
+  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
+    sequentialWakeUpRoutine().exitCode
+}
+```
+
+As we expected, the execution of the `run` show us that ZIO executes all of the effects in the same thread sequentially:
+
+```shell
+[zio-default-async-1]: Going to bath
+[zio-default-async-1]: Boiling some water
+[zio-default-async-1]: Preparing the coffee
+```
+
+## 4. Fibers
+
+If we want to leverage all the power of ZIO and starting to execute effects asynchronously, we must
+introduce the `Fiber` type.
+
+A _fiber_ is a concept that is beyond the ZIO library. In fact, it's a concurrency model. Often, we refer to fibers as _green threads_. **A fiber represents the blueprint of a running computation that may not have started yet**, like a thread. However, it's only a data type, which means **it doesn't map any operating system or JVM concept**. So its instantiation is very lightweight concerning a system thread. Hence, we can create millions of fibers and switch between them without the overheads associated with threads.
```

Comment:
I'd rephrase this paragraph a bit.

- "A fiber represents the blueprint of a running computation that may not have started yet, like a thread" -> "A fiber is a schedulable piece of computation, much like a thread" (the "runnable" and "may not have started yet" can spark confusions)
- data type -> data structure
- "doesn't map any operating system or JVM concept" -> "it's up to the ZIO runtime to schedule these fibers for execution (on the internal JVM thread pool)"
- "So its instantiation is very lightweight concerning a system thread" -> "unlike a system/JVM thread which is expensive to start and stop, fibers are cheap to allocate and remove".